### PR TITLE
[externalComponent] Add persistence support for external component

### PR DESCRIPTION
**Pull request by lolimay**
深度截图_选择区域_20190821221305

### CODE REVIEW

**Comment by d-gubert** [comment]
Hey @lolimay , apparently the file `externalComponent.js` has been removed? Is it intended?

File: app/apps/client/externalComponent/persistence.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import { Meteor } from 'meteor/meteor';
+
+export function setItem(appId, key, value) {
+	Meteor.call('externalComponentStorage:setItem', appId, key, value);
```
[the remark this answers is not in the document]
Comment (lolimay) [reply app/apps/client/externalComponent/persistence.js]: Got it. Will re-implement it with REST API.

**Comment by lolimay** [comment]
> Hey @lolimay, apparently the file `externalComponent.js` has been removed? Is it intended?

Yeah, I moved `externalComponent.js` into `app/apps/externalComponent` folder and change its name to `index.js`. I also add a new file `persistence.js` in this folder to keep all `externalCompoent` related files together. Any suggestions?

**Comment by d-gubert** [comment]
> Yeah, I moved `externalComponent.js` into `app/apps/externalComponent` folder and change its name to `index.js`. I also add a new file `persistence.js` in this folder to keep all `externalCompoent` related files together. Any suggestions?

Oh, I see! It's ok, then :)

File: app/apps/client/externalComponent/index.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
+import { Meteor } from 'meteor/meteor';
+import { Session } from 'meteor/session';
+import { Template } from 'meteor/templating';
+
+import { APIClient } from '../../../utils';
+import { getUserAvatarURL } from '../../../utils/lib/getUserAvatarURL';
+import { setItem, getItem, getAll, clear, removeItem } from './persistence';
+
+Meteor.startup(function() {
+	window.addEventListener('message', async ({ data, source }) => {
```
Comment (d-gubert) [inline app/apps/client/externalComponent/index.js]: Just a comment here, we can work on that later on.

I think this part of the code should live in the Apps-Engine, as a `AppsEngineUIHost` or something similar. Then, we can keep the communication logic in one repo and make sure they are compatible for every Apps-Engine version.

As it is currently implemented, apps would have not only a dependency to the Apps-Engine version, _but a dependency to Rocket.Chat's version as well_, and things will begin to get "messier", harder to maintain.

Of course, if we move this part into the Apps-Engine, we are going to need some dependencies to be injected in the `AppsEngineUIHost` class, as it won't have easy access to `Meteor` variable, for instance. But we are going to benefit from moving away from tight coupling to the `Meteor` framework, which is a big plus :)

**Comment by lolimay** [comment]
The related changes in the Rocket.Chat.Apps-engine repo is here: https://github.com/RocketChat/Rocket.Chat.Apps-engine/pull/135

**Pull request closed without merging**